### ISSUE TITLE
Bug using clogit with multiple strata

### ISSUE THREAD

**Issue by aridf** [vincentarelbundock/marginaleffects#946]
### Issue:

Calling `marginaleffects()` on a clogit model with multiple strata produces an error. Hopefully this is not just a versioning error on my part. In the meantime, I can just create a new variable to avoid having two variables in `strata`

### Reprex:
```
mtest <- clogit(am ~ mpg + cyl + mpg:cyl + strata(carb, gear), data = mtcars)
marginaleffects(mtest, newdata = mtcars, type = 'expected')

Error in str2lang(termtext) : <text>:1:5: unexpected ','
1: carb,
```

### SessionInfo output:

```
R version 4.2.0 (2022-04-22 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows Server x64 (build 20348)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.utf8  LC_CTYPE=English_United States.utf8   
[3] LC_MONETARY=English_United States.utf8 LC_NUMERIC=C                          
[5] LC_TIME=English_United States.utf8    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] marginaleffects_0.16.0 fstcore_0.9.14         dtplyr_1.1.0          
 [4] fst_0.9.8              survival_3.3-1         forcats_1.0.0         
 [7] stringr_1.4.0          dplyr_1.1.3            purrr_0.3.4           
[10] readr_1.4.0            tidyr_1.1.3            tibble_3.2.1          
[13] ggplot2_3.4.3          tidyverse_1.3.1       

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.11         lubridate_1.7.10    lattice_0.20-45     zoo_1.8-9          
 [5] assertthat_0.2.1    utf8_1.2.1          R6_2.5.0            cellranger_1.1.0   
 [9] backports_1.4.1     reprex_2.0.0        httr_1.4.2          pillar_1.9.0       
[13] rlang_1.1.1         readxl_1.3.1        rstudioapi_0.13     data.table_1.14.8  
[17] Matrix_1.4-1        checkmate_2.2.0     splines_4.2.0       munsell_0.5.0      
[21] broom_0.7.6         compiler_4.2.0      numDeriv_2016.8-1.1 modelr_0.1.8       
[25] pkgconfig_2.0.3     insight_0.19.6.2    tidyselect_1.2.0    fixest_0.11.1      
[29] fansi_0.5.0         crayon_1.4.1        dbplyr_2.1.1        withr_2.5.0        
[33] grid_4.2.0          nlme_3.1-157        jsonlite_1.7.2      gtable_0.3.0       
[37] lifecycle_1.0.3     DBI_1.1.1           magrittr_2.0.1      scales_1.2.1       
[41] cli_3.6.0           stringi_1.5.3       dreamerr_1.3.0      fs_1.5.0           
[45] xml2_1.3.2          ellipsis_0.3.2      generics_0.1.3      vctrs_0.6.3        
[49] sandwich_3.0-2      Formula_1.2-4       tools_4.2.0         glue_1.4.2         
[53] hms_1.0.0           parallel_4.2.0      colorspace_2.0-1    rvest_1.0.0        
[57] haven_2.4.1
```

**Comment by vincentarelbundock** [comment]
Thanks for the report. This is a problem upstream in the `insight` package. I have reported it here: https://github.com/easystats/insight/issues/821

Once it is fixed upstream and you install the updated version of `insight`, `marginaleffects` should work automatically.

BTW, you should use the `slopes()` function instead of `marginaleffects()`. They are equivalent, but the latter will eventually be deprecated.

**Comment by aridf** [comment]
Hi @vincentarelbundock, actually even with a single strata it still doesn't work, albeit with a different, more general error. I see support for `clogit` was added in version 0.7.0, so this should work, but it doesn't. I get the same error for my actual use case, which is a much larger more complicated formula/dataset

```
mtest <- clogit(am ~ mpg + cyl + mpg:cyl + strata(carb), data = mtcars)
slopes(mtest, newdata = mtcars, type = 'expected')

Error: Unable to compute predicted values with this model. This error can arise when
  `insight::get_data()` is unable to extract the dataset from the model object, or
  when the data frame was modified since fitting the model. You can try to supply a
  different dataset to the `newdata` argument.
  
  Bug Tracker: https://github.com/vincentarelbundock/marginaleffects/issues
```

Removing `newdata`:

```
mtest <- clogit(am ~ mpg + cyl + mpg:cyl + strata(carb), data = mtcars)
slopes(mtest, type = 'expected')

Error: Unable to compute predicted values with this model. This error can arise when
  `insight::get_data()` is unable to extract the dataset from the model object, or
  when the data frame was modified since fitting the model. You can try to supply a
  different dataset to the `newdata` argument.
  
  Bug Tracker: https://github.com/vincentarelbundock/marginaleffects/issues
```

**Comment by vincentarelbundock** [comment]
This is weird. Do you know what could explain this error?

``` r
# Error in Surv(rep(1, 32L), am): Time and status are different lengths
```

Here’s the dataset that `slopes()` tries to use internally for computation:

``` r
hi <- structure(list(mpg = c(21, 21, 22.8, 21.4, 18.7, 18.1, 14.3, 
24.4, 22.8, 19.2, 17.8, 16.4, 17.3, 15.2, 10.4, 10.4, 14.7, 32.4, 
30.4, 33.9, 21.5, 15.5, 15.2, 13.3, 19.2, 27.3, 26, 30.4, 15.8, 
19.7, 15, 21.4, 21.00235, 21.00235, 22.80235, 21.40235, 18.70235, 
18.10235, 14.30235, 24.40235, 22.80235, 19.20235, 17.80235, 16.40235, 
17.30235, 15.20235, 10.40235, 10.40235, 14.70235, 32.40235, 30.40235, 
33.90235, 21.50235, 15.50235, 15.20235, 13.30235, 19.20235, 27.30235, 
26.00235, 30.40235, 15.80235, 19.70235, 15.00235, 21.40235), 
    cyl = c(6.0004, 6.0004, 4.0004, 6.0004, 8.0004, 6.0004, 8.0004, 
    4.0004, 4.0004, 6.0004, 6.0004, 8.0004, 8.0004, 8.0004, 8.0004, 
    8.0004, 8.0004, 4.0004, 4.0004, 4.0004, 4.0004, 8.0004, 8.0004, 
    8.0004, 8.0004, 4.0004, 4.0004, 4.0004, 8.0004, 6.0004, 8.0004, 
    4.0004, 6, 6, 4, 6, 8, 6, 8, 4, 4, 6, 6, 8, 8, 8, 8, 8, 8, 
    4, 4, 4, 4, 8, 8, 8, 8, 4, 4, 4, 8, 6, 8, 4), disp = c(160, 
    160, 108, 258, 360, 225, 360, 146.7, 140.8, 167.6, 167.6, 
    275.8, 275.8, 275.8, 472, 460, 440, 78.7, 75.7, 71.1, 120.1, 
    318, 304, 350, 400, 79, 120.3, 95.1, 351, 145, 301, 121, 
    160, 160, 108, 258, 360, 225, 360, 146.7, 140.8, 167.6, 167.6, 
    275.8, 275.8, 275.8, 472, 460, 440, 78.7, 75.7, 71.1, 120.1, 
    318, 304, 350, 400, 79, 120.3, 95.1, 351, 145, 301, 121), 
    hp = c(110, 110, 93, 110, 175, 105, 245, 62, 95, 123, 123, 
    180, 180, 180, 205, 215, 230, 66, 52, 65, 97, 150, 150, 245, 
    175, 66, 91, 113, 264, 175, 335, 109, 110, 110, 93, 110, 
    175, 105, 245, 62, 95, 123, 123, 180, 180, 180, 205, 215, 
    230, 66, 52, 65, 97, 150, 150, 245, 175, 66, 91, 113, 264, 
    175, 335, 109), drat = c(3.9, 3.9, 3.85, 3.08, 3.15, 2.76, 
    3.21, 3.69, 3.92, 3.92, 3.92, 3.07, 3.07, 3.07, 2.93, 3, 
    3.23, 4.08, 4.93, 4.22, 3.7, 2.76, 3.15, 3.73, 3.08, 4.08, 
    4.43, 3.77, 4.22, 3.62, 3.54, 4.11, 3.9, 3.9, 3.85, 3.08, 
    3.15, 2.76, 3.21, 3.69, 3.92, 3.92, 3.92, 3.07, 3.07, 3.07, 
    2.93, 3, 3.23, 4.08, 4.93, 4.22, 3.7, 2.76, 3.15, 3.73, 3.08, 
    4.08, 4.43, 3.77, 4.22, 3.62, 3.54, 4.11), wt = c(2.62, 2.875, 
    2.32, 3.215, 3.44, 3.46, 3.57, 3.19, 3.15, 3.44, 3.44, 4.07, 
    3.73, 3.78, 5.25, 5.424, 5.345, 2.2, 1.615, 1.835, 2.465, 
    3.52, 3.435, 3.84, 3.845, 1.935, 2.14, 1.513, 3.17, 2.77, 
    3.57, 2.78, 2.62, 2.875, 2.32, 3.215, 3.44, 3.46, 3.57, 3.19, 
    3.15, 3.44, 3.44, 4.07, 3.73, 3.78, 5.25, 5.424, 5.345, 2.2, 
    1.615, 1.835, 2.465, 3.52, 3.435, 3.84, 3.845, 1.935, 2.14, 
    1.513, 3.17, 2.77, 3.57, 2.78), qsec = c(16.46, 17.02, 18.61, 
    19.44, 17.02, 20.22, 15.84, 20, 22.9, 18.3, 18.9, 17.4, 17.6, 
    18, 17.98, 17.82, 17.42, 19.47, 18.52, 19.9, 20.01, 16.87, 
    17.3, 15.41, 17.05, 18.9, 16.7, 16.9, 14.5, 15.5, 14.6, 18.6, 
    16.46, 17.02, 18.61, 19.44, 17.02, 20.22, 15.84, 20, 22.9, 
    18.3, 18.9, 17.4, 17.6, 18, 17.98, 17.82, 17.42, 19.47, 18.52, 
    19.9, 20.01, 16.87, 17.3, 15.41, 17.05, 18.9, 16.7, 16.9, 
    14.5, 15.5, 14.6, 18.6), vs = c(0, 0, 1, 1, 0, 1, 0, 1, 1, 
    1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 
    0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 
    0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1), am = c(1, 
    1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 
    0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 
    1, 1, 1, 1, 1, 1), gear = c(4, 4, 4, 3, 3, 3, 3, 4, 4, 4, 
    4, 3, 3, 3, 3, 3, 3, 4, 4, 4, 3, 3, 3, 3, 3, 4, 5, 5, 5, 
    5, 5, 4, 4, 4, 4, 3, 3, 3, 3, 4, 4, 4, 4, 3, 3, 3, 3, 3, 
    3, 4, 4, 4, 3, 3, 3, 3, 3, 4, 5, 5, 5, 5, 5, 4), carb = c(4, 
    4, 1, 1, 2, 1, 4, 2, 2, 4, 4, 3, 3, 3, 4, 4, 4, 1, 2, 1, 
    1, 2, 2, 4, 2, 1, 2, 2, 4, 6, 8, 2, 4, 4, 1, 1, 2, 1, 4, 
    2, 2, 4, 4, 3, 3, 3, 4, 4, 4, 1, 2, 1, 1, 2, 2, 4, 2, 1, 
    2, 2, 4, 6, 8, 2), rowid = c(1L, 2L, 3L, 4L, 5L, 6L, 7L, 
    8L, 9L, 10L, 11L, 12L, 13L, 14L, 15L, 16L, 17L, 18L, 19L, 
    20L, 21L, 22L, 23L, 24L, 25L, 26L, 27L, 28L, 29L, 30L, 31L, 
    32L, 1L, 2L, 3L, 4L, 5L, 6L, 7L, 8L, 9L, 10L, 11L, 12L, 13L, 
    14L, 15L, 16L, 17L, 18L, 19L, 20L, 21L, 22L, 23L, 24L, 25L, 
    26L, 27L, 28L, 29L, 30L, 31L, 32L), term = c("cyl", "cyl", 
    "cyl", "cyl", "cyl", "cyl", "cyl", "cyl", "cyl", "cyl", "cyl", 
    "cyl", "cyl", "cyl", "cyl", "cyl", "cyl", "cyl", "cyl", "cyl", 
    "cyl", "cyl", "cyl", "cyl", "cyl", "cyl", "cyl", "cyl", "cyl", 
    "cyl", "cyl", "cyl", "mpg", "mpg", "mpg", "mpg", "mpg", "mpg", 
    "mpg", "mpg", "mpg", "mpg", "mpg", "mpg", "mpg", "mpg", "mpg", 
    "mpg", "mpg", "mpg", "mpg", "mpg", "mpg", "mpg", "mpg", "mpg", 
    "mpg", "mpg", "mpg", "mpg", "mpg", "mpg", "mpg", "mpg"), 
    contrast = c("dY/dX", "dY/dX", "dY/dX", "dY/dX", "dY/dX", 
    "dY/dX", "dY/dX", "dY/dX", "dY/dX", "dY/dX", "dY/dX", "dY/dX", 
    "dY/dX", "dY/dX", "dY/dX", "dY/dX", "dY/dX", "dY/dX", "dY/dX", 
    "dY/dX", "dY/dX", "dY/dX", "dY/dX", "dY/dX", "dY/dX", "dY/dX", 
    "dY/dX", "dY/dX", "dY/dX", "dY/dX", "dY/dX", "dY/dX", "dY/dX", 
    "dY/dX", "dY/dX", "dY/dX", "dY/dX", "dY/dX", "dY/dX", "dY/dX", 
    "dY/dX", "dY/dX", "dY/dX", "dY/dX", "dY/dX", "dY/dX", "dY/dX", 
    "dY/dX", "dY/dX", "dY/dX", "dY/dX", "dY/dX", "dY/dX", "dY/dX", 
    "dY/dX", "dY/dX", "dY/dX", "dY/dX", "dY/dX", "dY/dX", "dY/dX", 
    "dY/dX", "dY/dX", "dY/dX")), row.names = c(NA, -64L), class = "data.frame")
```

I can use it with the basic `predict()` function without issues when `type="lp"`, but it breaks when `type="expected"`. Frankly, I don’t know this modelling package well enough to know what’s going on here. Do you have any ideas?

``` r
library(survival)
mtest <- clogit(am ~ mpg + cyl + mpg:cyl + strata(carb), data = mtcars)

predict(mtest, newdata = mtcars, type = "lp") |> head()
#         Mazda RX4     Mazda RX4 Wag        Datsun 710    Hornet 4 Drive 
#          2.219077          2.219077         -1.847810         -2.936551 
# Hornet Sportabout           Valiant 
#         -2.437470         -4.779259

predict(mtest, newdata = hi, type = "lp") |> head()
#         1         2         3         4         5         6 
#  2.219070  2.219070 -1.847871 -2.936570 -2.437408 -4.779178

predict(mtest, newdata = mtcars, type = "expected") |> head()
# [1] 1.051028e+00 1.051028e+00 1.122812e-03 3.779833e-04 8.983370e-04
# [6] 5.986801e-05

predict(mtest, newdata = hi, type = "expected") |> head()
# Error in Surv(rep(1, 32L), am): Time and status are different lengths
```

**Comment by aridf** [comment]
I'm not sure.

So `survival::clogit` uses the same predict function as `coxph` because it's implemented as a special case of a cox proportional hazard model where time is a constant. The `expected` and `survival` predictions from `coxph` both require estimating the baseline hazard (see [here](https://stats.stackexchange.com/questions/479597/help-undestanding-coxph-output), which might explain the need for the original data. Does `marginaleffects` have solutions for working with cox models that could also be applied here?

I'm not sure if this restriction always has to be the case for conditional logits. There is also the `bife` package (https://cran.r-project.org/web/packages/bife/vignettes/howto.html) that implements conditional logit more straightforwardly (though I haven't used it), and I also suspect one could just use `fixest::feglm` with `family = binomial(link = "logit")` with properly specified fixed effects, if it weren't for the "incidental parameters problem" that both `survival::clogit` and `bife::bife` correct for.

But I could be wrong about all of this. I will keep looking into it and post here with any updates

**Comment by vincentarelbundock** [comment]
OK, I think I found a hint: The data frame does not need to be identical; it just needs to have the same number of rows. The problem is that behind the scenes `slopes()` tries to compute all the slopes at the same time (for computational efficiency), so the prediction data is longer than the original 32 rows. In contrast, if you just do one variable at a time, it should work:

```r
slopes(mtest, variables = "mpg", type = 'expected')
```

I don't have a good solution for factor levels, since there we will typically have more rows than the original data...

**Comment by strengejacke** [comment]
Internally, the outcome seems to be converted to `Surv(rep(1, 32L)`, which might explain the issue (I guess this is what you have spotted, too)

**Comment by vincentarelbundock** [comment]
Thanks all. This is fixed upstream, so updating the `insight` package should fix this. 

The length of dataset error is much harder. I would argue that this is actually a design flaw in the `predict()` method offered by the `Survival` package. They should not simply do `rep(1, n)`. Instead, they probably should do something like `rep(1, nrow(newdata))`.

Unfortunately, this is outside of my scope to fix, so I will close this issue. At least we have documented the problem and there is a partial solution (even if inconvenient).
